File: src/maca/context.nim
## Context classes for managing different types of LLM interactions

import std/[json, httpclient, os, strutils, times, tables, strformat]
import logger, git_ops, utils
import tools  # Import tools module

type
  ContextError* = object of CatchableError

  Context* = ref object
    contextId*: string
    contextType*: string
    apiKey*: string
    messages*: seq[JsonNode]
    cumulativeCost*: int
    agentsMdContent*: string
    lastHeadCommit*: string
    defaultModel*: string
    model*: string
    toolNames*: seq[string]
    toolSchemas*: seq[JsonNode]
    logger*: Logger
    # References to maca instance data
    repoRoot*: string
    sessionId*: int
    worktreePath*: string

var instanceCounters: Table[string, int] = initTable[string, int]()

proc loadSystemPrompt(ctx: Context, promptsDir: string) =
  ## Load the system prompt from markdown files and parse metadata

  # First load common.md (shared across all contexts)
  let commonPath = promptsDir / "common.md"
  if not fileExists(commonPath):
    raise newException(ContextError, "Common prompt not found: " & commonPath)

  let commonPrompt = readFile(commonPath)
  ctx.messages.add(%*{
    "role": "system",
    "content": commonPrompt
  })

  # Then load context-specific prompt
  let promptPath = promptsDir / (ctx.contextType & ".md")
  if not fileExists(promptPath):
    raise newException(ContextError, "System prompt not found: " & promptPath)

  let promptContent = readFile(promptPath)

  # Split into headers and prompt body on first blank line
  let parts = promptContent.split("\n\n", 1)
  if parts.len != 2:
    raise newException(ContextError, "Prompt file must have headers separated by blank line")

  let headersText = parts[0]
  let systemPrompt = parts[1]

  # Parse headers
  for line in headersText.splitLines():
    let trimmed = line.strip()
    if trimmed == "":
      continue

    if ':' notin trimmed:
      raise newException(ContextError, "Invalid header format: " & line)

    let colonPos = trimmed.find(':')
    let key = trimmed[0 ..< colonPos].strip()
    let value = trimmed[colonPos + 1 .. ^1].strip()

    case key
    of "default_model":
      ctx.defaultModel = value
    of "tools":
      ctx.toolNames = value.split(',').mapIt(it.strip())
    else:
      raise newException(ContextError, "Unknown header key: " & key)

  ctx.messages.add(%*{
    "role": "system",
    "content": systemPrompt
  })

proc loadAgentsMd(ctx: Context) =
  ## Load AGENTS.md from the worktree if it exists
  if ctx.worktreePath == "":
    return

  let agentsPath = ctx.worktreePath / "AGENTS.md"
  if fileExists(agentsPath):
    let content = readFile(agentsPath)
    ctx.agentsMdContent = content
    ctx.messages.add(%*{
      "role": "system",
      "content": "# Project Context (AGENTS.md)\n\n" & content
    })

proc diffAgentsMd(ctx: Context) =
  ## Check if AGENTS.md has been updated and append diff to context
  if ctx.worktreePath == "":
    return

  let agentsPath = ctx.worktreePath / "AGENTS.md"
  if not fileExists(agentsPath):
    return

  let newContent = readFile(agentsPath)

  # Check if content has changed
  if newContent == ctx.agentsMdContent:
    return

  # For simplicity, just add a system message about the update
  # In a full implementation, you could compute an actual diff
  ctx.agentsMdContent = newContent
  ctx.messages.add(%*{
    "role": "system",
    "content": "# AGENTS.md Updated\n\nAGENTS.md has been updated with new content."
  })

proc checkHeadChanges(ctx: Context) =
  ## Check if HEAD has changed since last invocation
  if ctx.worktreePath == "" or ctx.lastHeadCommit == "":
    return

  let currentHead = getHeadCommit(cwd = ctx.worktreePath)

  if currentHead != ctx.lastHeadCommit:
    # HEAD has changed, gather info
    let commits = getCommitsBetween(ctx.lastHeadCommit, currentHead, cwd = ctx.worktreePath)
    let changedFiles = getChangedFilesBetween(ctx.lastHeadCommit, currentHead, cwd = ctx.worktreePath)

    if commits.len > 0 or changedFiles.len > 0:
      # Build system message
      var messageParts = @["# Repository Updates\n\nThe following changes have been made since you were last invoked:\n"]

      if commits.len > 0:
        messageParts.add("\n## New Commits\n")
        for commit in commits:
          messageParts.add(&"- `{commit.hash}` {commit.message}")

      if changedFiles.len > 0:
        messageParts.add("\n\n## Changed Files\n")
        for filepath in changedFiles:
          messageParts.add(&"- {filepath}")

      ctx.messages.add(%*{
        "role": "system",
        "content": messageParts.join("\n")
      })

    # Update tracking
    ctx.lastHeadCommit = currentHead

proc newContext*(
  contextType: string,
  repoRoot: string,
  sessionId: int,
  worktreePath: string,
  model = "auto",
  contextId = "",
  initialMessage = ""
): Context =
  ## Initialize a context
  ##
  ## Args:
  ##   contextType: Type of context (main, code_analysis, research, etc.)
  ##   repoRoot: Repository root path
  ##   sessionId: Session ID number
  ##   worktreePath: Path to the worktree
  ##   model: Model to use ("auto" to use default from prompt)
  ##   contextId: Optional context ID (auto-generated if not provided)
  ##   initialMessage: Optional initial user message

  var finalContextId = contextId
  if finalContextId == "":
    # Auto-generate unique name
    if not instanceCounters.hasKey(contextType):
      instanceCounters[contextType] = 0
    instanceCounters[contextType] += 1
    finalContextId = contextType & $instanceCounters[contextType]

  result = Context(
    contextId: finalContextId,
    contextType: contextType,
    repoRoot: repoRoot,
    sessionId: sessionId,
    worktreePath: worktreePath,
    messages: @[],
    cumulativeCost: 0,
    defaultModel: "openai/gpt-5-mini",
    toolNames: @[]
  )

  # Get API key from environment
  result.apiKey = getEnv("OPENROUTER_API_KEY")
  if result.apiKey == "":
    raise newException(ContextError, "OPENROUTER_API_KEY not set")

  # Initialize logger
  result.logger = newLogger(repoRoot, sessionId, finalContextId)

  # Load system prompt and parse metadata
  let scriptDir = getAppDir()
  let promptsDir = scriptDir / "prompts"
  result.loadSystemPrompt(promptsDir)

  # Get tool schemas
  result.toolSchemas = getToolSchemas(result.toolNames, addRationale = (contextType != "_main"))

  # Set model
  if model == "auto":
    result.model = result.defaultModel
  else:
    result.model = model

  # Load AGENTS.md if it exists
  result.loadAgentsMd()

  # Add unique name info
  result.messages.add(%*{
    "role": "system",
    "content": "Your unique context identifier is: **" & finalContextId & "**"
  })

  # Initialize HEAD tracking
  if worktreePath != "":
    result.lastHeadCommit = getHeadCommit(cwd = worktreePath)

  # Add initial message if provided
  if initialMessage != "":
    result.messages.add(%*{
      "role": "user",
      "content": initialMessage
    })

proc addMessage*(ctx: Context, message: JsonNode) =
  ## Add a message to the context and log it
  ctx.logger.log([("tag", %"message"), ("content", message)])
  ctx.messages.add(message)

proc callLlm*(ctx: Context): tuple[message: JsonNode, cost: int] =
  ## Call the LLM and return the response
  ##
  ## Returns tuple with message and cost in microdollars

  let client = newHttpClient()
  client.headers = newHttpHeaders({
    "Content-Type": "application/json",
    "Authorization": "Bearer " & ctx.apiKey,
    "HTTP-Referer": "https://github.com/vanviegen/maca",
    "X-Title": "MACA - Multi-Agent Coding Assistant"
  })

  let requestBody = %*{
    "model": ctx.model,
    "messages": ctx.messages,
    "tools": ctx.toolSchemas,
    "usage": {"include": true},
    "tool_choice": "required"
  }

  let startTime = epochTime()

  try:
    let response = client.postContent(
      "https://openrouter.ai/api/v1/chat/completions",
      body = $requestBody
    )

    let responseJson = parseJson(response)
    let choice = responseJson["choices"][0]
    let message = choice["message"]

    # Extract usage
    let usage = responseJson["usage"]
    let cost = int(usage["cost"].getFloat() * 1_000_000)  # Convert to microdollars
    ctx.cumulativeCost += cost

    let duration = epochTime() - startTime

    ctx.logger.log([
      ("tag", %"llm_call"),
      ("model", %ctx.model),
      ("cost", %cost),
      ("prompt_tokens", usage["prompt_tokens"]),
      ("completion_tokens", usage["completion_tokens"]),
      ("duration", %duration)
    ])

    # Add assistant message to history
    ctx.addMessage(message)

    return (message, cost)

  except Exception as e:
    raise newException(ContextError, "LLM API error: " & e.msg)
  finally:
    client.close()

proc run*(ctx: Context, budget = 0): tuple[summary: string, completed: bool, cost: int] =
  ## Run this context until completion or budget exceeded
  ##
  ## Args:
  ##   budget: Maximum cost in microdollars (0 = unlimited)
  ##
  ## Returns:
  ##   Tuple with summary, completion status, and total cost

  var totalCost = 0
  var completed = false
  var summaryParts: seq[string] = @[]
  let isSubcontext = ctx.contextType != "_main"
  let indent = if isSubcontext: "  " else: ""

  # Main loop
  while not completed:
    colorPrintLn(cyan(indent & "Context '" & ctx.contextId & "' thinking..."))

    # Check for AGENTS.md updates
    ctx.diffAgentsMd()

    # Check for HEAD changes
    ctx.checkHeadChanges()

    # Call LLM with retry logic
    var llmResult: tuple[message: JsonNode, cost: int]
    var success = false

    for attempt in 0..2:
      try:
        llmResult = ctx.callLlm()
        success = true
        break
      except Exception as e:
        colorPrintLn(red(indent & "Error during LLM call: " & e.msg & ". Retrying..."))
        summaryParts.add("Error during LLM call: " & e.msg)

    if not success:
      break

    let message = llmResult.message
    let cost = llmResult.cost
    totalCost += cost

    # Extract tool calls
    if not message.hasKey("tool_calls"):
      raise newException(ContextError, "No tool calls in response")

    let toolCalls = message["tool_calls"].getElems()
    if toolCalls.len != 1:
      raise newException(ContextError, "Expected exactly 1 tool call, got " & $toolCalls.len)

    let toolCall = toolCalls[0]
    let toolName = toolCall["function"]["name"].getStr()
    let toolArgs = parseJson(toolCall["function"]["arguments"].getStr())

    # Extract rationale if present
    var rationale = ""
    if toolArgs.hasKey("rationale"):
      rationale = toolArgs["rationale"].getStr()

    # Log tool call
    ctx.logger.log([("tag", %"tool_call"), ("tool", %toolName), ("args", toolArgs)])

    # Print tool info
    colorPrint(green(indent & "→"), plain(" Tool: "), yellow(toolName))
    if rationale != "":
      colorPrintLn(plain(""))
      colorPrintLn(plain(indent & "  Rationale: " & rationale))
    else:
      colorPrintLn(plain(""))

    # Execute tool
    let startTime = epochTime()
    var toolResult: JsonNode
    var toolError = false

    try:
      # Create tool context
      let toolCtx = ToolContext(
        worktreePath: ctx.worktreePath,
        repoRoot: ctx.repoRoot,
        sessionId: ctx.sessionId
      )

      toolResult = executeTool(toolName, toolArgs, toolCtx)

    except Exception as e:
      toolError = true
      toolResult = %*{"error": e.msg}
      colorPrintLn(red(indent & "Tool error: " & e.msg))

    let toolDuration = epochTime() - startTime

    # Check if this is a ready result
    var isReady = false
    if toolResult.hasKey("ready") and toolResult["ready"].getBool():
      isReady = true
      completed = true

    ctx.logger.log([
      ("tag", %"tool_result"),
      ("tool", %toolName),
      ("duration", %toolDuration),
      ("result", toolResult),
      ("completed", %completed)
    ])

    # Add tool result to messages
    ctx.messages.add(%*{
      "role": "tool",
      "tool_call_id": toolCall["id"],
      "content": $toolResult
    })

    # Check for git changes and commit if needed
    if not toolError and ctx.worktreePath != "":
      let diffStats = getDiffStats(ctx.worktreePath)
      if diffStats != "":
        # Commit changes
        let commitMsg = if rationale != "": &"{toolName}: {rationale}" else: toolName
        discard commitChanges(ctx.worktreePath, commitMsg)
        ctx.logger.log([("tag", %"commit"), ("message", %commitMsg), ("diff_stats", %diffStats)])
        colorPrintLn(green(indent & "✓ Committed changes"))

    # Build summary for this iteration
    var abbrArgs = $toolArgs
    if abbrArgs.len > 120:
      abbrArgs = abbrArgs[0..79] & "..."
    let iterationSummary = &"Called {toolName}({abbrArgs}) because: {rationale}"
    summaryParts.add(iterationSummary)

    if completed:
      colorPrintLn(green(indent & &"✓ Context {ctx.contextId} completed. Cost: {totalCost}μ$"))
      ctx.logger.log([("tag", %"complete")])
      if toolResult.hasKey("message"):
        summaryParts.add(toolResult["message"].getStr())
      break

    # Check budget (only for subcontexts)
    if budget > 0 and totalCost > budget:
      let budgetMsg = &"Context '{ctx.contextId}' budget exceeded (spent {totalCost}μ$ of {budget}μ$)"
      colorPrintLn(yellow(indent & budgetMsg))
      summaryParts.add(budgetMsg)
      break

  result = (summaryParts.join("\n"), completed, totalCost)

File: context.py
#!/usr/bin/env python3
"""Context class for managing LLM interactions."""

import time
from pathlib import Path
from typing import Dict, Any, Optional
import json
import os
import urllib.request


class ContextError(Exception):
    """Context operation failed."""
    pass


class Context:
    """Context for managing LLM interactions."""

    def __init__(
        self,
        model: str = "auto",
        initial_message: Optional[str] = None
    ):
        """
        Initialize a context.

        Args:
            model: Model to use for this context ("auto" to use default from prompt)
            initial_message: Optional initial user message to add to context
        """
        self.context_id = "main"
        self.api_key = os.environ.get('OPENROUTER_API_KEY')
        self._messages = []
        self.cumulative_cost = 0
        self.last_head_commit = None
        self.default_model = 'openai/gpt-5-mini'

        self.logger = Logger(maca.repo_root, maca.session_id, self.context_id)

        if not self.api_key:
            raise ContextError("OPENROUTER_API_KEY not set")

        # Load system prompt and parse metadata
        self.tool_names = []
        self._load_system_prompt()
        self.tool_schemas = tools.get_tool_schemas(self.tool_names, add_rationale=False)

        # Set model (use provided or default from prompt)
        if model == "auto":
            self.model = self.default_model
        else:
            self.model = model

        # Initialize HEAD tracking if we have a worktree
        self.last_head_commit = git_ops.get_head_commit(cwd=maca.worktree_path)

        # Add initial message if provided
        if initial_message:
            self.add_message({'role': 'user', 'content': initial_message})

    def _load_system_prompt(self):
        """Load the system prompt from prompt.md and parse metadata."""
        # Find prompt.md (next to the script)
        script_dir = Path(__file__).parent
        prompt_path = script_dir / 'prompt.md'

        if not prompt_path.exists():
            raise ContextError(f"System prompt not found: {prompt_path}")

        prompt_content = prompt_path.read_text()

        # Split into headers and prompt body on first blank line
        parts = prompt_content.split('\n\n', 1)
        if len(parts) != 2:
            raise ContextError(f"Prompt file {prompt_path} must have headers separated by blank line")

        headers_text, system_prompt = parts

        # Parse headers
        for line in headers_text.split('\n'):
            line = line.strip()
            if not line:
                continue

            if ':' not in line:
                raise ContextError(f"Invalid header format in {prompt_path}: {line}")

            key, value = line.split(':', 1)
            key = key.strip()
            value = value.strip()

            if key == 'default_model':
                self.default_model = value
            elif key == 'tools':
                self.tool_names = [name.strip() for name in value.split(',')]
            else:
                raise ContextError(f"Unknown header key in {prompt_path}: {key}")

        self.add_message({
            'role': 'system',
            'content': system_prompt
        })


    def _check_head_changes(self):
        """
        Check if HEAD has changed since last invocation.

        If changed, add a system message with commit info and changed files.
        """
        if not maca.worktree_path or not self.last_head_commit:
            return

        current_head = git_ops.get_head_commit(cwd=maca.worktree_path)

        if current_head != self.last_head_commit:
            # HEAD has changed, gather info
            commits = git_ops.get_commits_between(self.last_head_commit, current_head, cwd=maca.worktree_path)
            changed_files = git_ops.get_changed_files_between(self.last_head_commit, current_head, cwd=maca.worktree_path)

            if commits or changed_files:
                # Build system message
                message_parts = ["# Repository Updates\n\nThe following changes have been made since you were last invoked:\n"]

                if commits:
                    message_parts.append("\n## New Commits\n")
                    for commit in commits:
                        message_parts.append(f"- `{commit['hash']}` {commit['message']}")

                if changed_files:
                    message_parts.append("\n\n## Changed Files\n")
                    for filepath in changed_files:
                        message_parts.append(f"- {filepath}")

                self.add_message({
                    'role': 'system',
                    'content': '\n'.join(message_parts)
                })

            # Update tracking
            self.last_head_commit = current_head

    def call_llm(self) -> Dict[str, Any]:
        """
        Call the LLM and return the response.

        Returns:
            Dict with 'message' and 'tool_calls' keys
        """

        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {self.api_key}',
            'HTTP-Referer': 'https://github.com/vanviegen/maca',
            'X-Title': 'MACA - Multi-Agent Coding Assistant'
        }

        data = {
            'model': self.model,
            'messages': self._messages,
            'tools': self.tool_schemas,
            'usage': {"include": True},
            'tool_choice': 'required',  # Force tool use
        }
        start_time = time.time()

        try:
            req = urllib.request.Request(
                "https://openrouter.ai/api/v1/chat/completions",
                data=json.dumps(data).encode('utf-8'),
                headers=headers
            )
            with urllib.request.urlopen(req) as response:
                result = json.loads(response.read().decode('utf-8'))
        except Exception as e:
            if hasattr(e, 'read'):
                error_body = e.read().decode('utf-8')
                raise ContextError(f"LLM API error: {error_body}")
            raise ContextError(f"LLM API error: {str(e)}")

        # Extract response
        choice = result['choices'][0]
        message = choice['message']

        # Extract usage
        usage = result.get('usage', {})
        cost = int(usage.get('cost', 0) * 1_000_000)  # Convert dollars to microdollars
        self.cumulative_cost += cost

        self.logger.log(tag='llm_call', model=self.model, cost=cost, prompt_tokens=usage['prompt_tokens'], completion_tokens=usage['completion_tokens'], duration=time.time() - start_time)
        self.logger.log(tag='full_response', **result) # debugging only

        # Add assistant message to history
        self.add_message(message)

        return {
            'message': message,
            'cost': cost
        }
    
    def add_message(self, message: Dict):
        """Add a message dict to the context and the log."""
        self.logger.log(tag="message", **message)
        self._messages.append(message)

    def run(self):
        """
        Run this context until completion.

        Returns:
            None (runs until complete() is called)
        """
        completed = False

        # Loop until completion
        while not completed:
            # Print thinking message
            color_print(('ansicyan', "Thinking..."))

            # Check for HEAD changes before calling LLM
            self._check_head_changes()

            # Call LLM
            for _ in range(3):  # Retry up to 3 times
                try:
                    result = self.call_llm()
                    break
                except Exception as err:
                    color_print(('ansired', f"Error during LLM call: {err}. Retrying..."))
                    self.logger.log(tag='error', error=str(err))
            else:
                break

            # Extract tool calls
            message = result['message']
            cost = result['cost']

            # Extract tool info
            tool_calls = message.get('tool_calls', [])
            if len(tool_calls) != 1:
                raise ContextError(f"Expected exactly 1 tool call, got {len(tool_calls)}")
            tool_call = tool_calls[0]
            tool_name = tool_call['function']['name']
            tool_args = json.loads(tool_call['function']['arguments'])

            # Log tool call
            self.logger.log(tag='tool_call', tool=tool_name, args=str(tool_args))

            # Print tool info
            color_print(('ansigreen', '→'), ' Tool: ', ('ansiyellow', f"{tool_name}({tool_args})"))

            # Execute tool
            tool_start = time.time()
            try:
                immediate_result, context_summary = tools.execute_tool(tool_name, tool_args)
                tool_duration = time.time() - tool_start
            except Exception as err:
                immediate_result = {"error": str(err)}
                context_summary = f"{tool_name}: error"
                tool_duration = time.time() - tool_start

            # Check if this is a completion signal
            if isinstance(immediate_result, tools.ReadyResult):
                immediate_result = immediate_result.result
                completed = True

            self.logger.log(tag='tool_result', tool=tool_name, duration=tool_duration, result=immediate_result, completed=completed)

            # Convert immediate result to string for LLM
            immediate_output = immediate_result if isinstance(immediate_result, str) else json.dumps(immediate_result)

            # Check if output is long (>500 chars) - needs summarization
            if not completed and len(immediate_output) > 500:
                # Add ephemeral cache control marker before the long message
                self.add_message({
                    'role': 'user',
                    'content': '',
                    'cache_control': {'type': 'ephemeral'}
                })

                # Add the full immediate result (temporary, one-time view)
                self.add_message({
                    'type': 'function_call_output',
                    'call_id': tool_call['id'],
                    'output': immediate_output
                })

                color_print(('ansiyellow', f'→ Long output ({len(immediate_output)} chars), requesting summary...'))

                # Make LLM call that REQUIRES summarize_and_update_files
                summary_schemas = tools.get_tool_schemas(['summarize_and_update_files'], add_rationale=False)

                headers = {
                    'Content-Type': 'application/json',
                    'Authorization': f'Bearer {self.api_key}',
                    'HTTP-Referer': 'https://github.com/vanviegen/maca',
                    'X-Title': 'MACA - Coding Assistant'
                }

                data = {
                    'model': self.model,
                    'messages': self._messages,
                    'tools': summary_schemas,
                    'usage': {"include": True},
                    'tool_choice': {'type': 'function', 'function': {'name': 'summarize_and_update_files'}},
                }

                try:
                    req = urllib.request.Request(
                        "https://openrouter.ai/api/v1/chat/completions",
                        data=json.dumps(data).encode('utf-8'),
                        headers=headers
                    )
                    with urllib.request.urlopen(req) as response:
                        summary_result = json.loads(response.read().decode('utf-8'))

                    # Extract summary tool call
                    choice = summary_result['choices'][0]
                    summary_message = choice['message']
                    usage = summary_result.get('usage', {})
                    summary_cost = int(usage.get('cost', 0) * 1_000_000)
                    self.cumulative_cost += summary_cost

                    self.logger.log(tag='llm_call', model=self.model, cost=summary_cost,
                                  prompt_tokens=usage['prompt_tokens'], completion_tokens=usage['completion_tokens'])

                    # Add assistant message
                    self.add_message(summary_message)

                    # Execute the summarize_and_update_files tool
                    summary_tool_calls = summary_message.get('tool_calls', [])
                    if summary_tool_calls:
                        summary_tool_call = summary_tool_calls[0]
                        summary_tool_args = json.loads(summary_tool_call['function']['arguments'])

                        # Execute summarize_and_update_files
                        _, final_summary = tools.execute_tool('summarize_and_update_files', summary_tool_args)

                        # Remove the last 3 messages (cache control, long output, summary assistant message)
                        self._messages = self._messages[:-3]

                        # Add the final summary to permanent context
                        self.add_message({
                            'type': 'function_call_output',
                            'call_id': tool_call['id'],
                            'output': final_summary
                        })

                        color_print(('ansigreen', f'✓ Summarized to: {final_summary}'))
                    else:
                        # Fallback if no tool call
                        self._messages = self._messages[:-2]
                        self.add_message({
                            'type': 'function_call_output',
                            'call_id': tool_call['id'],
                            'output': context_summary
                        })

                except Exception as e:
                    color_print(('ansired', f'Error during summarization: {e}'))
                    # Fallback to context summary
                    self._messages = self._messages[:-2]
                    self.add_message({
                        'type': 'function_call_output',
                        'call_id': tool_call['id'],
                        'output': context_summary
                    })
            else:
                # Short output or completion - use context summary
                self.add_message({
                    'type': 'function_call_output',
                    'call_id': tool_call['id'],
                    'output': context_summary
                })

            # Check for git changes and commit if needed
            diff_stats = git_ops.get_diff_stats(maca.worktree_path)
            if diff_stats:
                # Commit changes
                commit_msg = tool_name
                git_ops.commit_changes(maca.worktree_path, commit_msg)
                self.logger.log(tag='commit', message=commit_msg, diff_stats=diff_stats)
                color_print(('ansigreen', '✓ Committed changes'))

            if completed:
                color_print(('ansigreen', f'✓ Task completed. Total cost: {self.cumulative_cost}μ$'))
                self.logger.log(tag='complete')
                break



from logger import Logger
from maca import maca
from utils import color_print
import tools
import git_ops
